src/chainstate/stacks/boot/cost-voting.clar

```diff
@@ -0,0 +1,205 @@
+        (cur-principal-votes (default-to u0 (get votes (map-get? principal-proposal-votes {
+            address: tx-sender,
+            proposal-id: proposal-id }))))
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err ERR_NO_SUCH_PROPOSAL))))
+    )
+
+    (asserts! (> amount u0) (err ERR_AMOUNT_NOT_POSITIVE))
+    (asserts! (< burn-block-height expiration-block-height) (err ERR_PROPOSAL_EXPIRED))
+    (asserts! (is-none (map-get? proposal-confirmed-id { proposal-id: proposal-id }))
+        (err ERR_PROPOSAL_CONFIRMED))
+
+    (unwrap! (stx-transfer? amount tx-sender (as-contract tx-sender)) (err ERR_INSUFFICIENT_FUNDS))
+    (unwrap! (ft-mint? cost-vote-token amount tx-sender) (err ERR_FT_MINT))
+
+    (map-set proposal-votes { proposal-id: proposal-id } { votes: (+ amount cur-votes) })
+    (map-set principal-proposal-votes { address: tx-sender, proposal-id: proposal-id}
+                                    { votes: (+ amount cur-principal-votes)})
+    (ok true))
+)
+
+;; Withdraw votes
+(define-public (withdraw-votes (proposal-id uint) (amount uint))
+    (let (
+        (cur-votes (default-to u0 (get votes (map-get? proposal-votes { proposal-id: proposal-id }))))
+        (cur-principal-votes (default-to u0 (get votes (map-get? principal-proposal-votes {
+            address: tx-sender,
+            proposal-id: proposal-id }))))
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err ERR_NO_SUCH_PROPOSAL))))
+        (sender tx-sender)
+    )
+
+    (asserts! (> amount u0) (err ERR_AMOUNT_NOT_POSITIVE))
+    (asserts! (>= cur-principal-votes amount) (err ERR_INSUFFICIENT_FUNDS))
+
+    (unwrap! (as-contract (stx-transfer? amount tx-sender sender)) (err ERR_STX_TRANSFER))
+    (unwrap! (as-contract (ft-transfer? cost-vote-token amount sender tx-sender))
+        (err ERR_FT_TRANSFER))
+
+    (map-set proposal-votes { proposal-id: proposal-id } { votes: (- cur-votes amount) })
+    (map-set principal-proposal-votes { address: tx-sender, proposal-id: proposal-id }
+                                        { votes: (- cur-principal-votes amount) })
+    (ok true))
+)
+
+;; Miner veto
+(define-public (veto (proposal-id uint))
+    (let (
+        (cur-vetos (default-to u0 (get vetos (map-get? proposal-vetos { proposal-id: proposal-id }))))
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err ERR_NO_SUCH_PROPOSAL))))
+        (vetoed (default-to false (get vetoed (map-get? exercised-veto { proposal-id: proposal-id,
+                                                                         veto-height: block-height }))))
+        (last-miner (unwrap! (get-block-info? miner-address (- block-height u1))
+            (err ERR_FETCHING_BLOCK_INFO)))
```

Comment:
@kantai I remember a conversation we had in the past about certain PoX-derived state, like the consensus hash, being unsafe to expose to the Clarity VM because it could cause the same block to have two different interpretations (https://github.com/blockstack/stacks-blockchain/issues/1805).  Does this concern still apply here, since a block's miner is derived from PoX state (i.e. which block wins sortition is determined in part from the PoX bitvector)?  I think my fix to #1805 removes this concern, but I just want to make sure.

Reply:
I don't think that concern applies to the winning block -- if a block is ever being evaluated, that means that block is the winner -- so data from that block is always safe to expose (e.g., who the miner is and how much they committed).